### ISSUE TITLE
Add options to not wrap long collections

### ISSUE THREAD

**Issue by andradei** [rust-lang/rustfmt#810]
An options, perhaps `ignore_arrays`, that would keep the following format:

```
let mut board: [[char; 3]; 3] = [
    ['O', 'O', 'O'],
    ['O', 'O', 'O'],
    ['O', 'O', 'O']
];
```

Instead of inlining everything like it does now:

`let mut board: [[char; 3]; 3] = [['O', 'O', 'O'], ['O', 'O', 'O'], ['O', 'O', 'O']];`

**Comment by nrc** [comment]
Are you after something specifically for arrays of arrays or for array literals in general?

**Comment by andradei** [comment]
Arrays of arrays (of vectors, or other collection type that can be represented that way).
Perhaps something that mimics the formatting of structs:

```
struct Point {
    x: i32,
    y: i32,
}
```

**Comment by Ryman** [comment]
Similar cases from https://github.com/exercism/xrust/pull/143.

```diff
-    run_test(&[
-        "   ",
-        "   ",
-        "   "
-    ]);
+    run_test(&["   ", "   ", "   "]);
```

```diff
-    let lines = vec![
-        "  +-+",
-        "  | |",
-        "+-+-+",
-        "| |  ",
-        "+-+  "
-        ];
+    let lines = vec!["  +-+", "  | |", "+-+-+", "| |  ", "+-+  "];
```

The reformatting really hurts readability for seeing the structure in the strings.

**Comment by ytmimi** [comment]
It's been a long time since this issue was opened . Confirming this is still reproducible with `rustfmt 1.5.1- (f2c31ba0 2022-07-19)` using the default configs. These days rustfmt has the [`array_width`](https://github.com/rust-lang/rustfmt/blob/master/Configurations.md#array_width) option which can be used to configure the max width of an array before it needs to wrap. Although it's not a perfect solution we can use `array_width` to get the desired formatting.

**Input**
```rust
fn main() {
    let mut board: [[char; 3]; 3] = [
        ['O', 'O', 'O'],
        ['O', 'O', 'O'],
        ['O', 'O', 'O'],
    ];
}
```

**output** `array_width=60` (Default)
```rust
fn main() {
    let mut board: [[char; 3]; 3] = [['O', 'O', 'O'], ['O', 'O', 'O'], ['O', 'O', 'O']];
}
```

**Output** `array_width=48` (highest value I could find that forces a wrap)
```rust
fn main() {
    let mut board: [[char; 3]; 3] = [
        ['O', 'O', 'O'],
        ['O', 'O', 'O'],
        ['O', 'O', 'O'],
    ];
}
```

**Comment by sourcefrog** [comment]
Although you can approximate it by specifying the width, it would be nice to express the intent more directly by saying that arrays containing compound members should have one member per line. (Though, I suppose this does not help the case with strings...)

**Comment by ytmimi** [comment]
Agreed that this isn't a direct solution, but just wanted to mention it. @sourcefrog would you mind providing a code snippet for the case you're envisioning.

**Comment by sourcefrog** [comment]
The first comment on this bug is an example of it. If it's an array (of length>1) of arrays, structs, or tuples (also non trivial) maybe there should be an entry per line.

**Comment by ytmimi** [comment]
Sorry, but I'm having trouble following. Could you please post the snippet that you had in mind. I think If I see it I might be able to understand it a little easier.

**Comment by sourcefrog** [comment]
instead of this

```
let mut board: [[char; 3]; 3] = [['O', 'O', 'O'], ['O', 'O', 'O'], ['O', 'O', 'O']];
```

it should produce this:

```
let mut board: [[char; 3]; 3] = [
    ['O', 'O', 'O'],
    ['O', 'O', 'O'],
    ['O', 'O', 'O']
];
```

However, if the elements of the array are scalars it should keep the current behavior of folding onto a single line, if it fits:

```
let mut board: [char; 9] = ['O', 'O', 'O', 'O', 'O', 'O', 'O', 'O', 'O'];
```

**Comment by calebcartwright** [comment]
@ytmimi - AIUI it's just reiterating the original ask from ~6.5 years ago even though there are existing width-based options that can approximate the end result.

To be honest, my inclination from a purely-trying-to-cleanup-the-backlog perspective is to close this given the age. I understand the ask and desire expressed by a couple folks, but this isn't something that we the rustfmt are ever going to work on, and no one else has taken a pass at in ~6.5 years either.

However, it probably wouldn't be too difficult to implement, nor for us to try to maintain, if someone else wanted to attempt to implement it.

I suspect you'd just need to incorporate logic to determine a separator tactic based on the contents at the following sites (assuming this is desired for slices too):

https://github.com/rust-lang/rustfmt/blob/f2c31ba04db6bf07c140c86bfbe2cb16f1b83a8e/src/patterns.rs#L249-L257

https://github.com/rust-lang/rustfmt/blob/f2c31ba04db6bf07c140c86bfbe2cb16f1b83a8e/src/expr.rs#L69-L77

Where I worry the complexity would come in would really be around trying to determine which types of contents warrant that classification, and of course thinking about it from a rustfmt perspective vs. a personal preference (i.e. how could rustfmt support different user preferences for what qualifies)
